core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HttpHealthChecker.java

```diff
@@ -89,35 +107,188 @@ private synchronized void check() {
+            subscription.request(1);
+            maybeSchedulePingCheck();
+        }
+
+        @Override
+        public void onNext(HttpObject obj) {
             if (closed) {
-                return null;
+                subscription.cancel();
+                return;
             }
 
-            boolean isHealthy = false;
-            if (res != null) {
-                switch (res.status().codeClass()) {
-                    case SUCCESS:
-                        maxLongPollingSeconds = getMaxLongPollingSeconds(res);
-                        isHealthy = true;
+            try {
+                if (!(obj instanceof ResponseHeaders)) {
+                    ReferenceCountUtil.release(obj);
+                    return;
+                }
+
+                final ResponseHeaders headers = (ResponseHeaders) obj;
+                updateLongPollingSettings(headers);
+
+                final HttpStatusClass statusClass = headers.status().codeClass();
+                switch (statusClass) {
+                    case INFORMATIONAL:
+                        maybeSchedulePingCheck();
                         break;
                     case SERVER_ERROR:
-                        maxLongPollingSeconds = getMaxLongPollingSeconds(res);
+                        receivedExpectedResponse = true;
+                        break;
+                    case SUCCESS:
+                        isHealthy = true;
+                        receivedExpectedResponse = true;
                         break;
                     default:
-                        if (res.status() == HttpStatus.NOT_MODIFIED) {
-                            maxLongPollingSeconds = getMaxLongPollingSeconds(res);
+                        if (headers.status() == HttpStatus.NOT_MODIFIED) {
                             isHealthy = wasHealthy;
+                            receivedExpectedResponse = true;
                         } else {
                             // Do not use long polling on an unexpected status for safety.
                             maxLongPollingSeconds = 0;
+                            logger.warn("{} Unexpected health check response: {}", reqCtx, headers);
                         }
                 }
-            } else {
+            } finally {
+                subscription.request(1);
+            }
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            updateHealth();
+        }
+
+        @Override
+        public void onComplete() {
+            updateHealth();
+        }
+
+        private void updateLongPollingSettings(ResponseHeaders headers) {
+            final String longPollingSettings = headers.get(ARMERIA_LPHC);
+            if (longPollingSettings == null) {
                 maxLongPollingSeconds = 0;
+                pingIntervalSeconds = 0;
+                return;
             }
 
+            final int commaPos = longPollingSettings.indexOf(',');
+            long maxLongPollingSeconds = 0;
```

Comment:
Don't think these need to be `long`, especially the ping interval

Reply:
All changed to `int`